field/src/bn254.rs

```diff
@@ -0,0 +1,265 @@
+use core::iter::{Product, Sum};
+use core::ops::{Add, AddAssign, Div, DivAssign, Mul, MulAssign, Neg, Sub, SubAssign};
+use itertools::Itertools;
+use serde::{Serialize, Deserialize};
+use num::bigint::BigUint;
+use num::{Integer, One};
+use rand::RngCore;
+use std::str::FromStr;
+
+use crate::types::{Field, Sample};
+use crate::types::PrimeField as native_pf;
+
+#[derive(Copy, Clone, Serialize, Deserialize)]
+pub struct Bn254Field(pub [u64; 4]);
+
+fn biguint_from_array(arr: [u64; 4]) -> BigUint {
+    BigUint::from_slice(&[
+        arr[0] as u32,
+        (arr[0] >> 32) as u32,
+        arr[1] as u32,
+        (arr[1] >> 32) as u32,
+        arr[2] as u32,
+        (arr[2] >> 32) as u32,
+        arr[3] as u32,
+        (arr[3] >> 32) as u32,
+    ])
+}
+
+impl Default for Bn254Field {
+    fn default() -> Self {
+        Self::ZERO
+    }
+}
+
+impl PartialEq for Bn254Field {
+    fn eq(&self, other: &Self) -> bool {
+        self.to_canonical_biguint() == other.to_canonical_biguint()
+    }
+}
+
+
+impl Eq for Bn254Field {}
+
+impl Hash for Bn254Field {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.to_canonical_biguint().hash(state)
+    }
+}
+
+impl Display for Bn254Field {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        Display::fmt(&self.to_canonical_biguint(), f)
+    }
+}
+
+impl Debug for Bn254Field {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        Debug::fmt(&self.to_canonical_biguint(), f)
+    }
+}
+
+impl Sample for Bn254Field {
+    #[inline]
+    fn sample<R>(rng: &mut R) -> Self
+        where
+            R: rand::RngCore + ?Sized,
+    {
+        use num::bigint::RandBigInt;
+        Self::from_noncanonical_biguint(rng.gen_biguint_below(&Self::order()))
+    }
+}
+
+impl Field for Bn254Field {
+    const ZERO: Self = Self([0; 4]);
+
+    const ONE: Self = Self([1, 0, 0, 0]);
+
+    const TWO: Self = Self([2, 0, 0, 0]);
+
+    const NEG_ONE: Self = Self([
+        4891460686036598785 as u64,
+        2896914383306846353 as u64,
+        13281191951274694749 as u64,
+        3486998266802970665 as u64
+    ]);
+
+    const TWO_ADICITY: usize = 28;
+
+    const CHARACTERISTIC_TWO_ADICITY: usize = Self::TWO_ADICITY;
+
+    const MULTIPLICATIVE_GROUP_GENERATOR: Self = Self([5, 0, 0, 0]);
+
+    const POWER_OF_TWO_GENERATOR: Self = Self([
+        268435456 as u64,
+        0,
+        0,
+        0
+    ]);
+
+    const BITS: usize = 254;
+
+    fn order() -> BigUint {
+        BigUint::from_str("21888242871839275222246405745257275088548364400416034343698204186575808495617").unwrap()
+    }
+
+    fn characteristic() -> BigUint {
+        Self::order()
+    }
+
+    fn try_inverse(&self) -> Option<Self> {
+        if self.is_zero() {
+            return None;
+        }
+
+        // Fermat's Little Theorem
+        Some(self.exp_biguint(&(Self::order() - BigUint::one() - BigUint::one())))
+    }
+
+    fn from_noncanonical_biguint(n: BigUint) -> Self {
```

Comment:
Taking a modulus here with the field order is necessary.

Reply:
done